analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -197,89 +197,95 @@ private boolean canCombineWithLastRegisteredActivity(ActivityPayload payload, Ac
+	private void addActivity(ActivityPayload payload, GoalDTO matchingGoal, ActivityCacheResult activityCacheResult)
 	{
-		Goal matchingGoalEntity = goalService.getGoalEntityForUserAnonymizedID(userAnonymized.getID(), matchingGoal.getID());
-		DayActivity updatedDayActivity = createNewActivity(dayActivityCacheResult.content, payload, userAnonymized,
-				matchingGoalEntity);
-		dayActivityRepository.save(updatedDayActivity);
-		if (dayActivityCacheResult.shouldUpdateCache())
+		Goal matchingGoalEntity = goalService.getGoalEntityForUserAnonymizedID(payload.userAnonymized.getID(),
+				matchingGoal.getID());
+		Activity addedActivity = createNewActivity(payload, matchingGoalEntity);
+		if (shouldUpdateCache(activityCacheResult, addedActivity))
 		{
-			cacheService.updateLastDayActivityForUser(updatedDayActivity);
+			cacheService.updateLastActivityForUser(payload.userAnonymized.getID(), matchingGoal.getID(),
+					ActivityDTO.createInstance(addedActivity));
 		}
 
 		if (matchingGoal.isNoGoGoal())
 		{
-			sendConflictMessageToAllDestinationsOfUser(payload, userAnonymized, updatedDayActivity.getLastActivity(),
-					matchingGoalEntity);
+			sendConflictMessageToAllDestinationsOfUser(payload, addedActivity, matchingGoalEntity);
 		}
 	}
 
-	private void updateActivityEndTime(ActivityPayload payload, UserAnonymizedDTO userAnonymized, GoalDTO matchingGoal,
-			DayActivityCacheResult dayActivity, Activity activity)
+	private void updateActivityEndTime(ActivityPayload payload, GoalDTO matchingGoal, ActivityCacheResult activityCacheResult)
 	{
-		assert userAnonymized.getID().equals(dayActivity.content.getUserAnonymized().getID());
-		assert matchingGoal.getID().equals(dayActivity.content.getGoal().getID());
-
+		DayActivity dayActivity = findExistingDayActivity(payload, matchingGoal.getID());
+		// TODO: are we sure that getLastActivity() gives the same activity? no concurrency issues?
+		Activity activity = dayActivity.getLastActivity();
 		activity.setEndTime(payload.endTime);
-		dayActivityRepository.save(dayActivity.content);
-		if (dayActivity.shouldUpdateCache())
+		DayActivity updatedDayActivity = dayActivityRepository.save(dayActivity);
+		// TODO: are we sure that getLastActivity() gives the same activity? no concurrency issues?
+		Activity updatedActivity = updatedDayActivity.getLastActivity();
+		if (shouldUpdateCache(activityCacheResult, updatedActivity))
+		{
+			cacheService.updateLastActivityForUser(payload.userAnonymized.getID(), matchingGoal.getID(),
+					ActivityDTO.createInstance(updatedActivity));
+		}
+	}
+
+	private boolean shouldUpdateCache(ActivityCacheResult activityCacheResult, Activity newOrUpdatedActivity)
+	{
+		if (activityCacheResult.content == null)
 		{
-			cacheService.updateLastDayActivityForUser(dayActivity.content);
+			return activityCacheResult.status != ActivityCacheResultStatus.CACHED_HAS_LATER_DAY;
 		}
+		return !activityCacheResult.content.getEndTime().isAfter(newOrUpdatedActivity.getEndTime());
```

Comment:
It should include equals. Unfortunately, there is no `isBeforeOrEqual` method here. I'll see what I can do else.